apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerDataReader.cs

```diff
@@ -267,44 +270,62 @@ public override bool NextResult()
         /// <param name="cancellationToken">A cancellation token to cancel the read. Cloud Spanner currently
         /// supports limited cancellation while advancing the read to the next row.</param>
         /// <returns>True if another row was read.</returns>
-        public override Task<bool> ReadAsync(CancellationToken cancellationToken)
-        {
-            return ExecuteHelper.WithErrorTranslationAndProfiling(
-                async () =>
+        public override Task<bool> ReadAsync(CancellationToken cancellationToken) =>
+            ExecuteHelper.WithErrorTranslationAndProfiling(async () =>
                 {
-                    if (_metadata == null)
-                    {
-                        await PopulateMetadataAsync(cancellationToken).ConfigureAwait(false);
-                    }
-                    _rowValid = false;
-                    _innerList.Clear();
-                    
-                    var first = await _resultSet.NextAsync(cancellationToken).ConfigureAwait(false);
-                    if (first == null)
+                    // Note: the using statement ensures that the timer tasks is closed appropriately; otherwise
+                    // we could flood the system with timers.
+                    using (var timeoutCts = new CancellationTokenSource(TimeSpan.FromSeconds(_readTimeoutSeconds)))
                     {
-                        // If this is the first thing we've tried to read, then we know there are no rows.
-                        if (_hasRows == null)
+                        var timeoutToken = timeoutCts.Token;
+                        var effectiveToken = !cancellationToken.CanBeCanceled ? timeoutToken
```

Comment:
`effectiveToken` isn't being disposed of; which it probably should be if it's a linked source.